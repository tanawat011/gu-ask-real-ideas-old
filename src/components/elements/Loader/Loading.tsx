import { LOADING_COLOR, LOADING_STYLE } from '@constants'
import { Loading3FloorStyle, LoadingDashStyle } from '@element/svg'

interface PropsLoading {
  color?: LOADING_COLOR
  loaderStyle?: LOADING_STYLE
  className?: string
}

export const Loading: React.FC<PropsLoading> = (props) => {
  const {
    loaderStyle = LOADING_STYLE.DEFAULT,
    color = LOADING_COLOR.DEFAULT,
    className = '',
  } = props

  const renderColor = (classColor: LOADING_COLOR) => {
    if (loaderStyle === LOADING_STYLE.TWO_PIPE) {
      switch (classColor) {
        case LOADING_COLOR.SKY:
          return 'border-t-nextui-blue-400 border-b-nextui-blue-400'
        case LOADING_COLOR.WHITE:
          return 'border-t-white border-b-white'

        case LOADING_COLOR.DEFAULT:
        default:
          return 'border-t-nextui-gray-400 border-b-nextui-gray-400'
      }
    }

    if (loaderStyle === LOADING_STYLE.THREE_FLOOR) {
      switch (classColor) {
        case LOADING_COLOR.SKY:
          return 'fill-nextui-blue-400'
        case LOADING_COLOR.WHITE:
          return 'fill-white'

        case LOADING_COLOR.DEFAULT:
        default:
          return 'fill-nextui-gray-400'
      }
    }

    if (loaderStyle === LOADING_STYLE.DASH) {
      switch (classColor) {
        case LOADING_COLOR.SKY:
          return 'stroke-nextui-blue-400'
        case LOADING_COLOR.WHITE:
          return 'stroke-white'

        case LOADING_COLOR.DEFAULT:
        default:
          return 'stroke-nextui-gray-400'
      }
    }

    switch (classColor) {
      case LOADING_COLOR.SKY:
        return 'border-t-nextui-blue-400'
      case LOADING_COLOR.WHITE:
        return 'border-t-white'

      case LOADING_COLOR.DEFAULT:
      default:
        return 'border-t-nextui-gray-400'
    }
  }

  const renderLoader = (classColor: string) => {
    const baseClass = 'border-transparent border-4 border-t-4'

    switch (loaderStyle) {
      case LOADING_STYLE.TWO_PIPE:
        return (
          <div
            className={`loader-second ${baseClass} border-b-4 ${renderColor(
              color,
            )} ${classColor}`}
          />
        )

      case LOADING_STYLE.DASH:
        return LoadingDashStyle(classColor)

      case LOADING_STYLE.THREE_FLOOR:
        return Loading3FloorStyle(classColor)

      case LOADING_STYLE.DEFAULT:
      default:
        return <div className={`loader ${baseClass} ${classColor}`} />
    }
  }

  const classColor = renderColor(color)
  const loader = renderLoader(classColor)

  return <div className={className}>{loader}</div>
}
