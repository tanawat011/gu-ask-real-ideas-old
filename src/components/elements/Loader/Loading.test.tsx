import { render } from '@testing-library/react'
import { Loading } from './Loading'
import { LOADING_COLOR, LOADING_STYLE } from '@constants'

describe('<Loading />', () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })

  it(`should render to match snapshot with default style no props`, async () => {
    expect.assertions(1)

    const { asFragment } = render(<Loading />)

    expect(asFragment()).toMatchSnapshot()
  })

  Object.keys(LOADING_STYLE).forEach((style) => {
    it(`should render to match snapshot with ${style.toLowerCase()} style`, async () => {
      expect.assertions(1)

      const { asFragment } = render(
        <Loading
          loaderStyle={LOADING_STYLE[style as keyof typeof LOADING_STYLE]}
        />,
      )

      expect(asFragment()).toMatchSnapshot()
    })

    Object.keys(LOADING_COLOR).forEach((color) => {
      it(`should render to match snapshot with ${style.toLowerCase()} style with ${color.toLowerCase()} color`, async () => {
        expect.assertions(1)

        const { asFragment } = render(
          <Loading
            loaderStyle={LOADING_STYLE[style as keyof typeof LOADING_STYLE]}
            color={LOADING_COLOR[color as keyof typeof LOADING_COLOR]}
          />,
        )

        expect(asFragment()).toMatchSnapshot()
      })
    })
  })
})
