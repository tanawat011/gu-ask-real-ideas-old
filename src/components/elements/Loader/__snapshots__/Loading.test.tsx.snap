// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Loading /> should render to match snapshot with dash style 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <svg
      class="loader-dash stroke-nextui-gray-400"
      viewBox="0 0 50 50"
    >
      <circle
        class="path"
        cx="25"
        cy="25"
        fill="none"
        r="20"
        stroke-width="4"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`<Loading /> should render to match snapshot with dash style with default color 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <svg
      class="loader-dash stroke-nextui-gray-400"
      viewBox="0 0 50 50"
    >
      <circle
        class="path"
        cx="25"
        cy="25"
        fill="none"
        r="20"
        stroke-width="4"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`<Loading /> should render to match snapshot with dash style with sky color 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <svg
      class="loader-dash stroke-nextui-blue-400"
      viewBox="0 0 50 50"
    >
      <circle
        class="path"
        cx="25"
        cy="25"
        fill="none"
        r="20"
        stroke-width="4"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`<Loading /> should render to match snapshot with dash style with white color 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <svg
      class="loader-dash stroke-white"
      viewBox="0 0 50 50"
    >
      <circle
        class="path"
        cx="25"
        cy="25"
        fill="none"
        r="20"
        stroke-width="4"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`<Loading /> should render to match snapshot with default style 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <div
      class="loader border-transparent border-4 border-t-4 border-t-nextui-gray-400"
    />
  </div>
</DocumentFragment>
`;

exports[`<Loading /> should render to match snapshot with default style no props 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <div
      class="loader border-transparent border-4 border-t-4 border-t-nextui-gray-400"
    />
  </div>
</DocumentFragment>
`;

exports[`<Loading /> should render to match snapshot with default style with default color 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <div
      class="loader border-transparent border-4 border-t-4 border-t-nextui-gray-400"
    />
  </div>
</DocumentFragment>
`;

exports[`<Loading /> should render to match snapshot with default style with sky color 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <div
      class="loader border-transparent border-4 border-t-4 border-t-nextui-blue-400"
    />
  </div>
</DocumentFragment>
`;

exports[`<Loading /> should render to match snapshot with default style with white color 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <div
      class="loader border-transparent border-4 border-t-4 border-t-white"
    />
  </div>
</DocumentFragment>
`;

exports[`<Loading /> should render to match snapshot with three_floor style 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <svg
      class="loader-3floor fill-nextui-gray-400"
      enable-background="new 0 0 100 100"
      id="L7"
      version="1.1"
      viewBox="0 0 100 100"
      x="0px"
      y="0px"
    >
      <path
        d="M31.6,3.5C5.9,13.6-6.6,42.7,3.5,68.4c10.1,25.7,39.2,38.3,64.9,28.1l-3.1-7.9c-21.3,8.4-45.4-2-53.8-23.3 c-8.4-21.3,2-45.4,23.3-53.8L31.6,3.5z"
      >
        <animatetransform
          attributeName="transform"
          attributeType="XML"
          dur="2s"
          from="0 50 50"
          repeatCount="indefinite"
          to="360 50 50"
          type="rotate"
        />
      </path>
      <path
        d="M42.3,39.6c5.7-4.3,13.9-3.1,18.1,2.7c4.3,5.7,3.1,13.9-2.7,18.1l4.1,5.5c8.8-6.5,10.6-19,4.1-27.7 c-6.5-8.8-19-10.6-27.7-4.1L42.3,39.6z"
      >
        <animatetransform
          attributeName="transform"
          attributeType="XML"
          dur="1s"
          from="0 50 50"
          repeatCount="indefinite"
          to="-360 50 50"
          type="rotate"
        />
      </path>
      <path
        d="M82,35.7C74.1,18,53.4,10.1,35.7,18S10.1,46.6,18,64.3l7.6-3.4c-6-13.5,0-29.3,13.5-35.3s29.3,0,35.3,13.5 L82,35.7z"
      >
        <animatetransform
          attributeName="transform"
          attributeType="XML"
          dur="2s"
          from="0 50 50"
          repeatCount="indefinite"
          to="360 50 50"
          type="rotate"
        />
      </path>
    </svg>
  </div>
</DocumentFragment>
`;

exports[`<Loading /> should render to match snapshot with three_floor style with default color 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <svg
      class="loader-3floor fill-nextui-gray-400"
      enable-background="new 0 0 100 100"
      id="L7"
      version="1.1"
      viewBox="0 0 100 100"
      x="0px"
      y="0px"
    >
      <path
        d="M31.6,3.5C5.9,13.6-6.6,42.7,3.5,68.4c10.1,25.7,39.2,38.3,64.9,28.1l-3.1-7.9c-21.3,8.4-45.4-2-53.8-23.3 c-8.4-21.3,2-45.4,23.3-53.8L31.6,3.5z"
      >
        <animatetransform
          attributeName="transform"
          attributeType="XML"
          dur="2s"
          from="0 50 50"
          repeatCount="indefinite"
          to="360 50 50"
          type="rotate"
        />
      </path>
      <path
        d="M42.3,39.6c5.7-4.3,13.9-3.1,18.1,2.7c4.3,5.7,3.1,13.9-2.7,18.1l4.1,5.5c8.8-6.5,10.6-19,4.1-27.7 c-6.5-8.8-19-10.6-27.7-4.1L42.3,39.6z"
      >
        <animatetransform
          attributeName="transform"
          attributeType="XML"
          dur="1s"
          from="0 50 50"
          repeatCount="indefinite"
          to="-360 50 50"
          type="rotate"
        />
      </path>
      <path
        d="M82,35.7C74.1,18,53.4,10.1,35.7,18S10.1,46.6,18,64.3l7.6-3.4c-6-13.5,0-29.3,13.5-35.3s29.3,0,35.3,13.5 L82,35.7z"
      >
        <animatetransform
          attributeName="transform"
          attributeType="XML"
          dur="2s"
          from="0 50 50"
          repeatCount="indefinite"
          to="360 50 50"
          type="rotate"
        />
      </path>
    </svg>
  </div>
</DocumentFragment>
`;

exports[`<Loading /> should render to match snapshot with three_floor style with sky color 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <svg
      class="loader-3floor fill-nextui-blue-400"
      enable-background="new 0 0 100 100"
      id="L7"
      version="1.1"
      viewBox="0 0 100 100"
      x="0px"
      y="0px"
    >
      <path
        d="M31.6,3.5C5.9,13.6-6.6,42.7,3.5,68.4c10.1,25.7,39.2,38.3,64.9,28.1l-3.1-7.9c-21.3,8.4-45.4-2-53.8-23.3 c-8.4-21.3,2-45.4,23.3-53.8L31.6,3.5z"
      >
        <animatetransform
          attributeName="transform"
          attributeType="XML"
          dur="2s"
          from="0 50 50"
          repeatCount="indefinite"
          to="360 50 50"
          type="rotate"
        />
      </path>
      <path
        d="M42.3,39.6c5.7-4.3,13.9-3.1,18.1,2.7c4.3,5.7,3.1,13.9-2.7,18.1l4.1,5.5c8.8-6.5,10.6-19,4.1-27.7 c-6.5-8.8-19-10.6-27.7-4.1L42.3,39.6z"
      >
        <animatetransform
          attributeName="transform"
          attributeType="XML"
          dur="1s"
          from="0 50 50"
          repeatCount="indefinite"
          to="-360 50 50"
          type="rotate"
        />
      </path>
      <path
        d="M82,35.7C74.1,18,53.4,10.1,35.7,18S10.1,46.6,18,64.3l7.6-3.4c-6-13.5,0-29.3,13.5-35.3s29.3,0,35.3,13.5 L82,35.7z"
      >
        <animatetransform
          attributeName="transform"
          attributeType="XML"
          dur="2s"
          from="0 50 50"
          repeatCount="indefinite"
          to="360 50 50"
          type="rotate"
        />
      </path>
    </svg>
  </div>
</DocumentFragment>
`;

exports[`<Loading /> should render to match snapshot with three_floor style with white color 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <svg
      class="loader-3floor fill-white"
      enable-background="new 0 0 100 100"
      id="L7"
      version="1.1"
      viewBox="0 0 100 100"
      x="0px"
      y="0px"
    >
      <path
        d="M31.6,3.5C5.9,13.6-6.6,42.7,3.5,68.4c10.1,25.7,39.2,38.3,64.9,28.1l-3.1-7.9c-21.3,8.4-45.4-2-53.8-23.3 c-8.4-21.3,2-45.4,23.3-53.8L31.6,3.5z"
      >
        <animatetransform
          attributeName="transform"
          attributeType="XML"
          dur="2s"
          from="0 50 50"
          repeatCount="indefinite"
          to="360 50 50"
          type="rotate"
        />
      </path>
      <path
        d="M42.3,39.6c5.7-4.3,13.9-3.1,18.1,2.7c4.3,5.7,3.1,13.9-2.7,18.1l4.1,5.5c8.8-6.5,10.6-19,4.1-27.7 c-6.5-8.8-19-10.6-27.7-4.1L42.3,39.6z"
      >
        <animatetransform
          attributeName="transform"
          attributeType="XML"
          dur="1s"
          from="0 50 50"
          repeatCount="indefinite"
          to="-360 50 50"
          type="rotate"
        />
      </path>
      <path
        d="M82,35.7C74.1,18,53.4,10.1,35.7,18S10.1,46.6,18,64.3l7.6-3.4c-6-13.5,0-29.3,13.5-35.3s29.3,0,35.3,13.5 L82,35.7z"
      >
        <animatetransform
          attributeName="transform"
          attributeType="XML"
          dur="2s"
          from="0 50 50"
          repeatCount="indefinite"
          to="360 50 50"
          type="rotate"
        />
      </path>
    </svg>
  </div>
</DocumentFragment>
`;

exports[`<Loading /> should render to match snapshot with two_pipe style 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <div
      class="loader-second border-transparent border-4 border-t-4 border-b-4 border-t-nextui-gray-400 border-b-nextui-gray-400 border-t-nextui-gray-400 border-b-nextui-gray-400"
    />
  </div>
</DocumentFragment>
`;

exports[`<Loading /> should render to match snapshot with two_pipe style with default color 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <div
      class="loader-second border-transparent border-4 border-t-4 border-b-4 border-t-nextui-gray-400 border-b-nextui-gray-400 border-t-nextui-gray-400 border-b-nextui-gray-400"
    />
  </div>
</DocumentFragment>
`;

exports[`<Loading /> should render to match snapshot with two_pipe style with sky color 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <div
      class="loader-second border-transparent border-4 border-t-4 border-b-4 border-t-nextui-blue-400 border-b-nextui-blue-400 border-t-nextui-blue-400 border-b-nextui-blue-400"
    />
  </div>
</DocumentFragment>
`;

exports[`<Loading /> should render to match snapshot with two_pipe style with white color 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <div
      class="loader-second border-transparent border-4 border-t-4 border-b-4 border-t-white border-b-white border-t-white border-b-white"
    />
  </div>
</DocumentFragment>
`;
